Remove built-in COM interop support from Mono runtime

jkoritzinsky opened the pull request:
This support is always disabled, so let's delete it.

Fixes #97703

jkoritzinsky committed:
Remove built-in COM interop support from Mono runtime

jkoritzinsky committed:
Merge bstr support into marshal.c, delete cominterop.h/c, remove remaining cominterop references in mono runtime.

jkoritzinsky committed:
Remove some more checks and make them asserts instead

jkoritzinsky committed:
Add prototype for mono_string_from_bstr

jkoritzinsky committed:
Remove unused variable

lambdageek commented on src/mono/mono/metadata/wrapper-types.h:
```diff
@@ -20,6 +20,4 @@ WRAPPER(STELEMREF, "stelemref")
 WRAPPER(UNBOX, "unbox")
 WRAPPER(WRITE_BARRIER, "write-barrier")
 WRAPPER(OTHER, "other")
-WRAPPER(COMINTEROP_INVOKE, "cominterop-invoke")
-WRAPPER(COMINTEROP, "cominterop")
```
@vargaz Do the wrapper types this leak out into public APIs anywhere? Do we need to keep the numbering the same?  ie, should these be replaced by

```c
WRAPPER(UNUSED1, "unused1")
WRAPPER(UNUSED2, "unused2")
```

jkoritzinsky replied:
I looked through the public API and there's no way to get at the wrapper type from the embedding API surface.

lambdageek reviewed (approved):
Thanks!

jkoritzinsky committed:
Inline finalizer check into Wasm-only branch

jkoritzinsky commented:
Looks like I need to make the AOT compiler resilient to encountering a `[ComImport]` type. I'll take a look tomorrow at that.

jkoritzinsky committed:
Don't fail the AOT compiler when encountering a COM type

lambdageek commented on src/mono/mono/metadata/class-init.c:
```diff
@@ -564,6 +564,11 @@ mono_class_create_from_typedef (MonoImage *image, guint32 type_token, MonoError
 		}
 	}
 
+	if (MONO_CLASS_IS_IMPORT (klass)) {
+		mono_class_set_failure_and_error (klass, error, "Built-in COM interop is not supported on the Mono runtime.");
+		goto parent_failure;
+	}
```
Apparently this is breaking some dependency injection tests. I guess we have to be lazier - maybe we need  to keep `m_class_is_com_object` and just throw a TLE if we try to instantiate it.

jkoritzinsky committed:
Block constructing RCW types at construction time.

We need to allow ComImport types to be loaded as ComImport interfaces can also be implemented entirely in managed code, but we can block objects of ComImport-based types from being created.

jkoritzinsky committed:
Merge branch 'main' of github.com:dotnet/runtime into no-mono-com

jkoritzinsky committed:
Emit PNSE for COMImport wrapper generation

jkoritzinsky committed:
Move check

jkoritzinsky committed:
Fix null handle return

jkoritzinsky committed:
Let the regular flow emit the error.

lambdageek commented:
Linux arm64 libraries test crash is possibly related:

```
Thread 11 (Thread 0xffffb2e620e0 (LWP 34) ".NET Long Runni"):
#0  0x0000ffffbd926800 in __GI___wait4 (pid=pid@entry=35, stat_loc=stat_loc@entry=0xffffb2e5d450, options=0, options@entry=-1293561008, usage=usage@entry=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:30
#1  0x0000ffffbd926758 in __GI___waitpid (pid=pid@entry=35, stat_loc=stat_loc@entry=0xffffb2e5d450, options=options@entry=-1293561008) at ./posix/waitpid.c:38
#2  0x0000ffffbd4c147c in dump_native_stacktrace (signal=<optimized out>, mctx=<optimized out>) at /__w/1/s/src/mono/mono/mini/mini-posix.c:843
#3  mono_dump_native_crash_info (signal=<optimized out>, mctx=mctx@entry=0xffffb2e5deb0, info=<optimized out>) at /__w/1/s/src/mono/mono/mini/mini-posix.c:870
#4  0x0000ffffbd472218 in mono_handle_native_crash (signal=0xffffbd31027b "SIGSEGV", mctx=0xffffb2e5deb0, info=0xffffb2e5e210) at /__w/1/s/src/mono/mono/mini/mini-exceptions.c:2967
#5  0x0000ffffbd3dfb1c in mono_sigsegv_signal_handler_debug (_dummy=11, _info=0xffffb2e5e210, context=0xffffb2e5e290, debug_fault_addr=0x20) at /__w/1/s/src/mono/mono/mini/mini-runtime.c:3888
#6  <signal handler called>
#7  mono_object_handle_isinst (obj=..., klass=klass@entry=0x0, error=error@entry=0xffffb2e5f4e0) at /__w/1/s/src/mono/mono/metadata/object.c:6807
#8  0x0000ffffbd5f212c in mono_object_isinst_checked (obj_raw=0xffffbca2a1e0, klass=0x0, error=error@entry=0xffffb2e5f4e0) at /__w/1/s/src/mono/mono/metadata/object.c:6790
#9  0x0000ffffbd5c2ee4 in mono_marshal_isinst_with_cache (obj=0xffffbca2a1e0, klass=0x0, cache=0xffff9005a820) at /__w/1/s/src/mono/mono/metadata/marshal.c:4445
#10 0x0000ffffbcd6ff7c in ?? ()
#11 0x0000ffff94003580 in ?? ()
```

Note `mono_marshal_isinst_with_cache is getting called from JITed code with a NULL `klass` argument.

jkoritzinsky commented:
@lambdageek There's a known issue for that failure #90019.

Looking at [one of the logs from one of the other instances](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-707c8ce34f1b49f08a/Microsoft.Extensions.Options.Tests/1/console.c304fa18.log?helixlogtype=result), it looks like the same symptom

```
#6  <signal handler called>
#7  mono_object_handle_isinst (obj=..., klass=0x0, error=0xf15fb198) at /__w/1/s/src/mono/mono/metadata/object.c:6884
#8  0xf77991e8 in mono_object_isinst_checked (obj_raw=0xf6a6ece0, klass=0x0, error=0xf15fb198) at /__w/1/s/src/mono/mono/metadata/object.c:6867
#9  0xf7763eb8 in mono_marshal_isinst_with_cache (obj=0xf6a6ece0, klass=0x0, cache=0xf1728fb0) at /__w/1/s/src/mono/mono/metadata/marshal.c:4356
#10 0xf6df0220 in ?? ()
```

As you're more familiar with this code than I, I'll defer to you.

lambdageek commented:
Oh yes, that's the same crash.  I didn't recognize the test suite. I think of 90019 as somehow being related to generic collections (perhaps some kind of generic sharing race condition).  But yes, it's the same issue.

:shipit:
